pkg/querier/tenantfederation/merge_queryable.go

```diff
@@ -144,23 +147,70 @@ type stringSliceFunc func(storage.Querier) ([]string, storage.Warnings, error)
 // the output of the stringSliceFunc to be sorted, as results of LabelValues
 // are not sorted.
 //
-// TODO: Consider running stringSliceFunc calls concurrently
+
+type stringSliceFuncResult struct {
+	Result   []string
+	Warnings storage.Warnings
+}
+
 func (m *mergeQuerier) mergeDistinctStringSlice(f stringSliceFunc) ([]string, storage.Warnings, error) {
+
+	g, _ := errgroup.WithContext(m.ctx)
+
 	var warnings storage.Warnings
 	resultMap := make(map[string]struct{})
-	for pos, tenantID := range m.tenantIDs {
-		result, resultWarnings, err := f(m.queriers[pos])
-		if err != nil {
-			return nil, nil, err
-		}
-		for _, e := range result {
-			resultMap[e] = struct{}{}
-		}
-		for _, w := range resultWarnings {
-			warnings = append(warnings, fmt.Errorf("error querying tenant id %s: %w", tenantID, w))
+	resultCh := make(chan *stringSliceFuncResult)
+	resultDone := make(chan struct{})
+
+	go func() {
+		defer func() {
+			close(resultDone)
+		}()
+		for result := range resultCh {
+			for _, e := range result.Result {
+				resultMap[e] = struct{}{}
+			}
+			warnings = append(warnings, result.Warnings...)
 		}
+	}()
+
+	for pos := range m.tenantIDs {
+		querier := m.queriers[pos]
+		tenantID := m.tenantIDs[pos]
+		g.Go(func() error {
+			result, resultWarnings, err := f(querier)
```

Comment:
Is there any timeout associated with the `f()` function? Just wondering how this will handle a very slow call to storage.

Reply:
That's something that currently needs to be handled right now in the `f()` method. I think having a context being passed for those would be incredibly helpful, as discussed in another thread.